core/store/src/trie/mem/arena/mod.rs

```diff
@@ -0,0 +1,259 @@
+mod alloc;
+use self::alloc::Allocator;
+use borsh::{BorshDeserialize, BorshSerialize};
+use mmap_rs::{MmapFlags, MmapMut, MmapOptions};
+use std::fmt::{Debug, Formatter};
+use std::hash::Hash;
+use std::mem::size_of;
+
+/// Arena to store in-memory trie nodes.
+/// Includes a bump allocator that can free nodes.
+///
+/// To allocate, deallocate, or mutate any allocated memory, a mutable
+/// reference to the `Arena` is needed.
+pub struct Arena {
+    memory: ArenaMemory,
+    allocator: Allocator,
+}
+
+/// Mmap-ed memory to host the in-memory trie nodes.
+/// A mutable reference to `ArenaMemory` can be used to mutate allocated
+/// memory, but not to allocate or deallocate memory.
+pub struct ArenaMemory {
+    mmap: MmapMut,
+}
+
+impl ArenaMemory {
+    fn new(max_size_in_bytes: usize) -> Self {
+        let mmap = MmapOptions::new(max_size_in_bytes)
+            .unwrap()
+            // We need NO_RESERVE so that the memory region doesn't actually
+            // take up space before the memory is used.
```

Comment:
Do we actually want to no_reserve? I’d argue we should leave that to the default OS-wide settings. This way real validators can disable overcommit and be sure nearcore won’t be killed because they forgot to add enough memory. Besides this only reserves swap space according to man, so it’s probably not a big issue even for non-production validators anyway.